caikit/core/model_manager.py

```diff
@@ -137,150 +142,153 @@ def _load_from_dir(self, module_path, load_singleton, *args, **kwargs):
-
-        # If configured backends is empty, add LOCAL backend to it
-        local_enabled = backend_types.LOCAL in configured_backends
-
-        # NOTE: Local backend can be disabled via `config.yml` but its enabled
-        # by default
-
-        log.debug2("Local enabled? %s", local_enabled)
-        module_implementations = MODULE_BACKEND_REGISTRY.get(
-            module_id, {backend_types.LOCAL: None} if local_enabled else {}
-        )
-        log.debug2(
-            "Number of available backend implementations found: %d",
-            len(module_implementations),
-        )
-
-        error.value_check(
-            "<COR84094427E>",
-            len(module_implementations) > 0,
-            "No implementation of {} available. You may need to `pip install {}`",
-            module_config.module_id,
-            self.get_module_class_from_config(module_config),
-        )
-
-        loaded_artifact = None
-        # instantiate object and return to user
-        log.debug("Loading the artifact!")
-
-        # Go through each configured backend in priority order and look for an
-        # implementation that matches
-        # NOTE: backend priority can be configured while configuring backend
-        for backend in configured_backends:
-            # NOTE: what if we have multiple backends available,
-            # this will currently only return 1st one of those
-            # based on priority
-            backend_impl_obj = module_implementations.get(backend)
-            # NOTE: LOCAL may not be marked as a specific implementation, thus skip this
-            # or if no implementation available of the provided backend, continue
-            if backend == backend_types.LOCAL or backend_impl_obj is None:
-                continue
-
-            backend_impl = backend_impl_obj.impl_class
-
-            # A particular model can be supported by multiple backends, i.e
-            # a model trained on LOCAL backend might be able to get loaded with
-            # Spark or Ray backend implementations of the same block.
-            # Here, we will try to get the list of supported load backends for
-            # each of the backend_types and check if they contain the
-            # backend type of the current model
-            supported_load_backends = self._get_supported_load_backends(backend_impl)
-
-            # Check if the module actually supports this backend for load
-            if model_backend in supported_load_backends:
-                log.debug(
-                    "%s backend implementation found for backend [%s]",
-                    backend_impl.__name__,
-                    backend,
+        # If this is a singleton load, the entire body of this function needs to
+        # be locked to avoid concurrent loads on the same model. Otherwise, we
+        # can freely load in parallel.
+        with self.singleton_lock(load_singleton):
+
+            # Using the module_path as a key, look for an instance preloaded in the
```

Comment:
I know we talked about reviving the `unique_id` content hash for the case where the same content is being loaded from multiple paths. Just noting that here for posterity